File: packages/protocol/src/extensions.ts
/**
 * Extension-related types.
 */

/** Type of extension determining its runtime behavior. */
export type ExtensionType = "runner" | "integration" | "command";

/** Runtime environment for integration extensions. */
export type ExtensionRuntime = "wasm" | "deno";

/** Configuration for runner-type extensions. */
export interface RunnerConfig {
  /** Command to execute */
  command: string;
  /** Arguments to pass (supports {{prompt}} template) */
  args: string[];
  /** Whether the runner is interactive */
  interactive?: boolean;
  /** Whether to use PTY */
  pty?: boolean;
}

/** A command exposed by an extension. */
export interface ExtensionCommand {
  /** Unique identifier for this command within the extension */
  id: string;
  /** Human-readable name */
  name: string;
  /** Optional description */
  description?: string;
  /** Keywords for search/filtering */
  keywords?: string[];
}

/** Network permission for integration extensions. */
export interface NetworkPermission {
  /** Allowed hosts */
  network: string[];
}

/** Permission types for extensions. */
export type Permission = string | NetworkPermission;

/** Extension manifest loaded from manifest.json. */
export interface ExtensionManifest {
  /** Unique identifier for the extension */
  id: string;
  /** Human-readable name */
  name: string;
  /** Version string (semver) */
  version: string;
  /** Type of extension */
  type: ExtensionType;
  /** Optional description */
  description?: string;
  /** Runner configuration (for runner type) */
  runner?: RunnerConfig;
  /** Runtime environment (for integration type) */
  runtime?: ExtensionRuntime;
  /** Main entry point (for integration type) */
  main?: string;
  /** Script to run (for command type) */
  script?: string;
  /** Commands exposed by this extension */
  commands?: ExtensionCommand[];
  /** Required permissions */
  permissions?: Permission[];
}

/** Represents a loaded extension at runtime. */
export interface Extension {
  /** Unique runtime ID */
  id: string;
  /** The extension manifest */
  manifest: ExtensionManifest;
  /** Path to the extension directory */
  path: string;
  /** Whether the extension is currently enabled */
  enabled: boolean;
}

/** Status of an extension. */
export type ExtensionStatus = "ready" | "running" | "error" | "disabled";

File: crates/protocol/src/extensions.rs
//! Extension-related types.

use serde::{Deserialize, Serialize};
use uuid::Uuid;

/// Type of extension determining its runtime behavior.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ExtensionType {
    /// Spawns and manages CLI processes with PTY support
    Runner,
    /// Runs in sandboxed WASM/Deno runtime with network access
    Integration,
    /// Simple scripts that run and return results
    Command,
}

/// Runtime environment for integration extensions.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ExtensionRuntime {
    Wasm,
    Deno,
}

/// Configuration for runner-type extensions.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RunnerConfig {
    /// Command to execute
    pub command: String,
    /// Arguments to pass (supports {{prompt}} template)
    pub args: Vec<String>,
    /// Whether the runner is interactive
    #[serde(default)]
    pub interactive: bool,
    /// Whether to use PTY
    #[serde(default)]
    pub pty: bool,
}

/// A command exposed by an extension.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExtensionCommand {
    /// Unique identifier for this command within the extension
    pub id: String,
    /// Human-readable name
    pub name: String,
    /// Optional description
    #[serde(default)]
    pub description: Option<String>,
    /// Keywords for search/filtering
    #[serde(default)]
    pub keywords: Vec<String>,
}

/// Network permission for integration extensions.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NetworkPermission {
    /// Allowed hosts
    pub network: Vec<String>,
}

/// Permission types for extensions.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum Permission {
    /// Simple permission string (e.g., "process:spawn", "pty")
    Simple(String),
    /// Network permission with allowed hosts
    Network(NetworkPermission),
}

/// Extension manifest loaded from manifest.json.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExtensionManifest {
    /// Unique identifier for the extension
    pub id: String,
    /// Human-readable name
    pub name: String,
    /// Version string (semver)
    pub version: String,
    /// Type of extension
    #[serde(rename = "type")]
    pub extension_type: ExtensionType,
    /// Optional description
    #[serde(default)]
    pub description: Option<String>,
    /// Runner configuration (for runner type)
    #[serde(default)]
    pub runner: Option<RunnerConfig>,
    /// Runtime environment (for integration type)
    #[serde(default)]
    pub runtime: Option<ExtensionRuntime>,
    /// Main entry point (for integration type)
    #[serde(default)]
    pub main: Option<String>,
    /// Script to run (for command type)
    #[serde(default)]
    pub script: Option<String>,
    /// Commands exposed by this extension
    #[serde(default)]
    pub commands: Vec<ExtensionCommand>,
    /// Required permissions
    #[serde(default)]
    pub permissions: Vec<Permission>,
}

/// Represents a loaded extension at runtime.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Extension {
    /// Unique runtime ID
    pub id: Uuid,
    /// The extension manifest
    pub manifest: ExtensionManifest,
    /// Path to the extension directory
    pub path: String,
    /// Whether the extension is currently enabled
    pub enabled: bool,
}

/// Status of an extension.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ExtensionStatus {
    /// Extension is loaded and ready
    Ready,
    /// Extension is currently running
    Running,
    /// Extension encountered an error
    Error,
    /// Extension is disabled
    Disabled,
}
